### PR TITLE
Update cursor api

### PR DESCRIPTION
**Pull request by alschmiedt**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves
1. Couldn't easily create a new cursor. 
2. We weren't using a different cursor for the flyout and the workspace
3. Bug - Fixes toggle on readonly ws 
4. Bug - Should be able to focus toolbox from anywhere

### Proposed Changes
1. Separate the cursor and cursor_svg object so that the cursor is solely in charge of the logic and the cursor_svg is solely in charge of drawing the cursor. 

2. Blockly navigation always gets the cursor from the main workspace or from the flyout workspace.

3. The flyout has a specific kind of cursor that can only go next and previous. 

4. A marker is a type of cursor that can not move. - This might need to be separated into its own class. A gnosis issue has been added.

### Test Coverage

Tested on:

### CODE REVIEW

File: core/keyboard_nav/cursor.js [hunk 1]
```diff
@@ -29,39 +29,38 @@ goog.provide('Blockly.Cursor');
...
-  /**
-   * Whether or not the cursor is a marker.
-   * @type {boolean} True if the cursor is a marker. False otherwise.
-   * @private
-   */
-  this.isMarker_ = !!opt_marker;
+/**
+ * The object in charge of drawing the visual representation of the current node.
+ * @type {Blockly.CursorSvg}
+ */
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor.js]: Any reason for this not to be private?

File: core/keyboard_nav/cursor.js [hunk 2]
```diff
@@ -29,39 +29,38 @@ goog.provide('Blockly.Cursor');
...
+Blockly.Cursor.prototype.drawer = null;
+
+/**
+ * Sets the object in charge of drawing the cursor.
+ * @param{Blockly.CursorSvg} drawer The object in charge of drawing the cursor.
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor.js]: Extra space after @param.

File: core/keyboard_nav/flyout_cursor.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+'use strict';
+
+goog.provide('Blockly.FlyoutCursor');
+
+goog.require('Blockly.Cursor');
+
+/**
+ * Class for a flyout cursor.
+ * This controls how a user navigates blocks in the flyout.
+ * @constructor
```
Comment (samelhusseini) [inline core/keyboard_nav/flyout_cursor.js]: @extends {Blockly.Cursor}

File: core/keyboard_nav/cursor_svg.js [hunk 4]
```diff
@@ -26,25 +26,20 @@
...
+ *     red one.
  * @constructor
  */
-Blockly.CursorSvg = function(workspace, opt_marker) {
-  Blockly.CursorSvg.superClass_.constructor.call(this, opt_marker);
+Blockly.CursorSvg = function(workspace, isMarker) {
   this.workspace_ = workspace;
   this.constants = new Blockly.blockRendering.ConstantProvider();
   this.constants.init();
+  this.isMarker_ = isMarker;
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor_svg.js]: This needs an annotation.  so does the workspace_ above it.

File: core/keyboard_nav/flyout_cursor.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    if (newNode) {
+      this.setLocation(newNode);
+    }
+    return newNode;
+  }
+};
+
+/**
+ * This is a  no-op since a flyout cursor can not go out.
+ * @return {null} null
```
Comment (samelhusseini) [inline core/keyboard_nav/flyout_cursor.js]: Full stop after null. 
```
@return {null} Always null.
```

File: core/keyboard_nav/flyout_cursor.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    if (newNode) {
+      this.setLocation(newNode);
+    }
+    return newNode;
+  }
+};
+
+/**
+ * This is a  no-op since a flyout cursor can not go in.
+ * @return {null} null
```
Comment (samelhusseini) [inline core/keyboard_nav/flyout_cursor.js]: Full stop after null. 
```
@return {null} Always null.
```

File: core/keyboard_nav/marker_cursor.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview The class representing a cursor.
+ * Used primarily for keyboard navigation.
```
Comment (samelhusseini) [inline core/keyboard_nav/marker_cursor.js]: @author annotation.

File: core/keyboard_nav/marker_cursor.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+ */
+'use strict';
+
+goog.provide('Blockly.MarkerCursor');
+
+goog.require('Blockly.Cursor');
+
```
Comment (samelhusseini) [inline core/keyboard_nav/marker_cursor.js]: Extra space after require.

File: core/keyboard_nav/flyout_cursor.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * @fileoverview The class representing a cursor.
+ * Used primarily for keyboard navigation.
+ */
+'use strict';
+
+goog.provide('Blockly.FlyoutCursor');
+
+goog.require('Blockly.Cursor');
+
```
Comment (samelhusseini) [inline core/keyboard_nav/flyout_cursor.js]: Extra space after require.

File: core/keyboard_nav/cursor_svg.js [hunk 10]
```diff
@@ -26,25 +26,20 @@
 
 goog.provide('Blockly.CursorSvg');
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor_svg.js]: 2 spaces between the provide / requires and start of the file.

File: core/workspace.js [hunk 11]
```diff
@@ -162,19 +154,37 @@ Blockly.Workspace.prototype.MAX_UNDO = 1024;
...
 
 /**
- * Adds cursor for keyboard navigation.
- * @return {!Blockly.Cursor} Cursor for keyboard navigation.
+ * Sets the cursor for keyboard navigation.
+ * @param {Blockly.Cursor} cursor The cursor used to navigate around the blockly
+ *     ast for keyboard navigation.
+ */
+Blockly.Workspace.prototype.setCursor = function(cursor) {
+  this.cursor = cursor;
```
Comment (samelhusseini) [inline core/workspace.js]: is this.cursor defined anywhere? Might want to keep it in the constructor and set it to null, so you can keep its annotation.

Comment (samelhusseini) [reply core/workspace.js]: Also, should it be private?

File: core/workspace.js [hunk 12]
```diff
@@ -162,19 +154,37 @@ Blockly.Workspace.prototype.MAX_UNDO = 1024;
...
+};
+
+/**
+ * Sets the marker for keyboard navigation.
+ * @param {Blockly.MarkerCursor} marker The marker used to mark a location for
+ *     keyboard navigation.
+ */
+Blockly.Workspace.prototype.setMarker = function(marker) {
+  this.marker = marker;
```
Comment (samelhusseini) [inline core/workspace.js]: Same as cursor, this.marker needs to be defined in the constructor, and potentially private.

File: core/keyboard_nav/cursor.js (hunk 1 above)
Comment (alschmiedt) [reply core/keyboard_nav/cursor.js]: done

File: core/keyboard_nav/cursor_svg.js (hunk 10 above)
Comment (alschmiedt) [reply core/keyboard_nav/cursor_svg.js]: done

File: core/keyboard_nav/cursor_svg.js (hunk 4 above)
Comment (alschmiedt) [reply core/keyboard_nav/cursor_svg.js]: Why just the workspace and marker and not the constants?

File: core/keyboard_nav/flyout_cursor.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview The class representing a cursor.
+ * Used primarily for keyboard navigation.
```
[the remark this answers is not in the document]
Comment (alschmiedt) [reply core/keyboard_nav/flyout_cursor.js]: done

File: core/keyboard_nav/flyout_cursor.js (hunk 9 above)
Comment (alschmiedt) [reply core/keyboard_nav/flyout_cursor.js]: done

File: core/keyboard_nav/flyout_cursor.js (hunk 3 above)
Comment (alschmiedt) [reply core/keyboard_nav/flyout_cursor.js]: done

File: core/keyboard_nav/flyout_cursor.js (hunk 6 above)
Comment (alschmiedt) [reply core/keyboard_nav/flyout_cursor.js]: done

File: core/keyboard_nav/flyout_cursor.js (hunk 5 above)
Comment (alschmiedt) [reply core/keyboard_nav/flyout_cursor.js]: done

File: core/keyboard_nav/marker_cursor.js (hunk 7 above)
Comment (alschmiedt) [reply core/keyboard_nav/marker_cursor.js]: done

File: core/workspace.js (hunk 11 above)
Comment (alschmiedt) [reply core/workspace.js]: done

File: core/workspace.js (hunk 12 above)
Comment (alschmiedt) [reply core/workspace.js]: done

File: core/keyboard_nav/cursor_svg.js [hunk 14]
```diff
@@ -26,19 +26,31 @@
...
+ *     cursor is not a marker.
  * @constructor
  */
-Blockly.CursorSvg = function(workspace, isMarker) {
+Blockly.CursorSvg = function(workspace, opt_marker) {
+  /**
+   * @type {!Blockly.Workspace}
+   * @private
+   */
+
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor_svg.js]: nit: no need for the extra line here

File: core/keyboard_nav/cursor_svg.js (hunk 4 above)
Comment (samelhusseini) [reply core/keyboard_nav/cursor_svg.js]: Sure constants as well.

File: core/keyboard_nav/cursor.js [hunk 15]
```diff
@@ -29,39 +30,39 @@ goog.provide('Blockly.Cursor');
...
-   * @type {boolean} True if the cursor is a marker. False otherwise.
-   * @private
-   */
-  this.isMarker_ = !!opt_marker;
+/**
+ * The object in charge of drawing the visual representation of the current node.
+ * @type {Blockly.CursorSvg}
+ * @private
+ */
+Blockly.Cursor.prototype.drawer_ = null;
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor.js]: Can you move this to the constructor. According to the Google Style guide all properties should be defined in the constructor (there's a browser performance reason for that).

File: core/keyboard_nav/cursor_svg.js [hunk 16]
```diff
@@ -26,25 +26,31 @@
...
  */
 Blockly.CursorSvg = function(workspace, opt_marker) {
-  Blockly.CursorSvg.superClass_.constructor.call(this, opt_marker);
+  /**
+   * @type {!Blockly.Workspace}
+   * @private
+   */
   this.workspace_ = workspace;
+
+  /**
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor_svg.js]: Description.

File: core/keyboard_nav/cursor_svg.js [hunk 17]
```diff
@@ -26,25 +26,31 @@
...
  * @param {boolean=} opt_marker True if the cursor is a marker. A marker is used
  *     to save a location and is an immovable cursor. False or undefined if the
  *     cursor is not a marker.
- * @extends {Blockly.Cursor}
  * @constructor
  */
 Blockly.CursorSvg = function(workspace, opt_marker) {
-  Blockly.CursorSvg.superClass_.constructor.call(this, opt_marker);
+  /**
+   * @type {!Blockly.Workspace}
```
Comment (samelhusseini) [inline core/keyboard_nav/cursor_svg.js]: Description.

File: core/workspace.js [hunk 18]
```diff
@@ -162,19 +168,37 @@ Blockly.Workspace.prototype.MAX_UNDO = 1024;
 Blockly.Workspace.prototype.connectionDBList = null;
 
 /**
- * Adds cursor for keyboard navigation.
- * @return {!Blockly.Cursor} Cursor for keyboard navigation.
+ * Sets the cursor for keyboard navigation.
+ * @param {Blockly.Cursor} cursor The cursor used to navigate around the blockly
+ *     ast for keyboard navigation.
```
Comment (samelhusseini) [inline core/workspace.js]: capitalize ast.

Comment (samelhusseini) [reply core/workspace.js]: and blockly

File: core/flyout_base.js [hunk 19]
```diff
@@ -56,6 +56,8 @@ Blockly.Flyout = function(workspaceOptions) {
    */
   this.workspace_ = new Blockly.WorkspaceSvg(workspaceOptions);
   this.workspace_.isFlyout = true;
+  this.workspace_.setCursor(new Blockly.FlyoutCursor());
```
Comment (samelhusseini) [inline core/flyout_base.js]: need to require these in this file.

**Pull request merged**